### PR TITLE
Add MD 0D

### PR DESCRIPTION
**Pull request by ElmiraShamlou**
## Fixes/Resolves:

(replace this with the issue # fixed or resolved, if no issue exists then a brief statement of what this PR does)

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by adam-a-a** [comment]
@bknueven - adding you to this review since it includes a new MD_channel_base similar to your rework related to RO and membrane_channel control volume.

File: watertap/unit_models/MD/MD_channel_0D.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                    b.dP_dx[t, x] * b.length / b.nfe for x in b.length_domain
+                )
+
+        else:
+            raise ConfigurationError(
+                f"Unrecognized pressure_change_type {pressure_change_type}"
+            )
+
+    def set_config(self, config):
+        self.config = config
```
Comment (adam-a-a) [inline watertap/unit_models/MD/MD_channel_0D.py]: Unsure why this is necessary- should it be removed?

Comment (ElmiraShamlou) [reply watertap/unit_models/MD/MD_channel_0D.py]: Yes. It should be removed.

File: watertap/unit_models/MD/MD_channel_0D.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            self.properties_out.initialize(
+                state_args=state_args_properties_out,
+                outlvl=outlvl,
+                optarg=optarg,
+                solver=solver,
+            )
+        else:
+            raise ConfigurationError(
+                "Either hot_ch or cold_ch must be set in the configuration."
+            )
```
Comment (adam-a-a) [inline watertap/unit_models/MD/MD_channel_0D.py:221]: Why is it one or the other? Shouldn't both be initialized? Also, it looks like you would need pass strings to `type` here, which I don't think we want to do.

Comment (ElmiraShamlou) [reply watertap/unit_models/MD/MD_channel_0D.py:221]: cold outlet and hot outlet are initialized differently based on inlet

File: watertap/unit_models/MD/membrane_distillation_0D.py [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+        "has_holdup",
+        ConfigValue(
+            default=False,
+            domain=In([False]),
+            description="Holdup construction flag - must be False",
+            doc="""Indicates whether holdup terms should be constructed or not.
+    **default** - False. Membrane units do not have defined volume, thus
+    this must be False.""",
+        ),
+    )
```
Comment (adam-a-a) [inline watertap/unit_models/MD/membrane_distillation_0D.py:78]: Any reason that we need `dynamic` and `has_holdup` in the main config block as well as in the config template for hot and cold channels? I would think you could eliminate these 2 from config_template, but maybe there is some reason for doing this.

File: watertap/unit_models/MD/membrane_distillation_0D.py [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+            implicit=True,
+            description="Arguments to use for constructing property packages",
+            doc="""A ConfigBlock with arguments to be passed to a property block(s)
+        and used when constructing these,
+        **default** - None.
+        **Valid values:** {
+        see property package for documentation.}""",
+        ),
+    )
+
```
Comment (adam-a-a) [inline watertap/unit_models/MD/membrane_distillation_0D.py:353]: Is there any need to have `property_package_vapor` specified for each channel, or can we just specify it once on the main `CONFIG` block?

Comment (adam-a-a) [reply watertap/unit_models/MD/membrane_distillation_0D.py:353]: From the tests, I see you need to specify twice (once for each channel). Maybe there is some alternative configuration where you might want to consider 2 different vapor packages that I am not seeing, but from my current perspective, I don't see any immediate use cases for having 2 different vapor property packages.

File: watertap/unit_models/MD/membrane_distillation_base.py [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+            )
+
+        @self.Expression(
+            self.flowsheet().config.time,
+            doc="Average cold side enthalpy flux expression",
+        )
+        def flux_enth_cold_avg(b, t):
+            return (
+                sum(b.flux_enth_cold[t, x] for x in self.difference_elements) / self.nfe
+            )
```
Comment (adam-a-a) [inline watertap/unit_models/MD/membrane_distillation_base.py:344]: Shouldn't these be consolidated and put in MD_channel_base (or MD_channel_0D)?

File: watertap/unit_models/MD/membrane_distillation_base.py [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+        )
+
+        self.flux_enth_cold = Var(
+            self.flowsheet().config.time,
+            self.difference_elements,
+            initialize=1e3,
+            bounds=(1e-4, 1e5),
+            units=pyunits.J * pyunits.s**-1 * pyunits.m**-2,
+            doc="cold side condensation enthalpy flux",
+        )
```
Comment (adam-a-a) [inline watertap/unit_models/MD/membrane_distillation_base.py]: Can't these go in one of the channel files? Then, you could later reference `hot_ch.flux_enth` or `cold_ch.flux_enth`

File: watertap/unit_models/MD/membrane_distillation_0D.py [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+    CONFIG.declare("hot_ch", _CONFIG_Template(doc="hot channel config arguments"))
+    CONFIG.declare("cold_ch", _CONFIG_Template(doc="cold channel config arguments"))
```
Comment (adam-a-a) [inline watertap/unit_models/MD/membrane_distillation_0D.py:355]: Minor preference here, but I think it'd be better to fully spell out `hot_channel` and `cold_channel`.

**Comment by lbianchi-lbl** [comment]
@ElmiraShamlou is planning to work on some details and documentation but overall should be soon ready for review.

**Review by TimBartholomew — APPROVED** [review]
I primarily reviewed the documentation and test file and it looks good.

One suggestion I have that would be for a later PR would be to change "ch" to "channel". Renaming won't be that big of deal, but lots of lines of code would change.

File: watertap/unit_models/MD/membrane_distillation_base.py (hunk 5 above)
Comment (ElmiraShamlou) [reply watertap/unit_models/MD/membrane_distillation_base.py:344]: Mass and heat fluxes connect the channels but are not defined for each individual channel. Even the eq_flux_hot_enth and eq_flux_cold_enth are defined by the mass flux that connects the hot and cold channels.

File: watertap/unit_models/MD/membrane_distillation_base.py (hunk 6 above)
Comment (ElmiraShamlou) [reply watertap/unit_models/MD/membrane_distillation_base.py]: Only these two variables, but not their associated constraints, may go to MD_channel_base.

File: watertap/unit_models/MD/membrane_distillation_0D.py (hunk 7 above)
Comment (ElmiraShamlou) [reply watertap/unit_models/MD/membrane_distillation_0D.py:355]: We can do this in future PRs as it requires changing many lines of code.

File: watertap/unit_models/MD/MD_channel_0D.py (hunk 1 above)
Comment (ElmiraShamlou) [reply watertap/unit_models/MD/MD_channel_0D.py]: removed

File: watertap/unit_models/MD/MD_channel_0D.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    def add_state_blocks(
+        self,
+        has_phase_equilibrium=None,
+        flow_direction=None,
+        property_package_vapor=None,
+        property_package_args_vapor=None,
+    ):
+        super().add_state_blocks(has_phase_equilibrium=has_phase_equilibrium)
+
+        # quack like a 1D model
```
Comment (savannahsakhai) [inline watertap/unit_models/MD/MD_channel_0D.py:58]: What is meant here? I think there is a typo.

File: watertap/unit_models/MD/MD_channel_base.py [hunk 9]
```diff
@@ -0,0 +1,794 @@
...
+            doc="Concentration polarization modulus",
+        )
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.length_domain,
+            solute_set,
+            doc="Concentration polarization modulus",
+        )
+        def eq_cp_modulus(b, t, x, j):  # pylint: disable=function-redefined
```
Comment (savannahsakhai) [inline watertap/unit_models/MD/MD_channel_base.py:294]: Was this comment meant to be left in?

File: watertap/unit_models/MD/membrane_distillation_base.py [hunk 10]
```diff
@@ -0,0 +1,741 @@
...
+        self.add_inlet_port(name="cold_ch_inlet", block=self.cold_ch)
+        self.add_outlet_port(name="cold_ch_outlet", block=self.cold_ch)
+
+        self._add_heat_flux()
+        self._add_mass_flux()
+
+        self.recovery_mass = Var(
+            self.flowsheet().config.time,
+            # self.config.hot_ch.property_package.phase_list,
+            # self.config.hot_ch.property_package.component_list,
```
Comment (savannahsakhai) [inline watertap/unit_models/MD/membrane_distillation_base.py:153]: Are these meant to be left this way? Are they serving a purpose?

File: watertap/unit_models/MD/membrane_distillation_base.py [hunk 11]
```diff
@@ -0,0 +1,741 @@
...
+        )
+
+        self.flux_enth_cold = Var(
+            self.flowsheet().config.time,
+            self.difference_elements,
+            initialize=1e3,
+            bounds=(1e-10, 1e5),
+            units=pyunits.J * pyunits.s**-1 * pyunits.m**-2,
+            doc="cold side condensation enthalpy flux",
+        )
```
Comment (savannahsakhai) [inline watertap/unit_models/MD/membrane_distillation_base.py:256]: The variables directly above this use units_meta. Is there a reason that these variables do not follow the same convention? I believe this only occurs a few times across the files.

File: watertap/unit_models/tests/test_membrane_distillation_0D.py [hunk 12]
```diff
@@ -0,0 +1,1624 @@
...
+            hot_ch_flow_mass * hot_ch_mass_frac_H2O
+        )
+
+        m.fs.unit.hot_ch_inlet.pressure[0].fix(hot_ch_pressure)
+        m.fs.unit.hot_ch_inlet.temperature[0].fix(273.15 + 90)
+        m.fs.unit.area.fix(membrane_area)
+        m.fs.unit.permeability_coef.fix(1e-10)
+        m.fs.unit.membrane_thickness.fix(1e-4)
+        m.fs.unit.membrane_tc.fix(0.2)
+        # m.fs.unit.hot_ch.cp_modulus.fix(1)
```
Comment (savannahsakhai) [inline watertap/unit_models/tests/test_membrane_distillation_0D.py:527]: Is this comment meant to be left this way?

**Review by savannahsakhai — APPROVED** [review]
I attempted to look through all of the files. It looks good to me.

I made comments on a few possible typos.

File: watertap/unit_models/MD/membrane_distillation_base.py (hunk 10 above)
Comment (adam-a-a) [reply watertap/unit_models/MD/membrane_distillation_base.py:153]: We typically delete commented code, so we can delete in a subsequent PR.

**Review by adam-a-a — APPROVED** [review]
LGTM - @ElmiraShamlou please create an issue tracking any comments here that we would consider in a follow-up PR.

File: watertap/unit_models/MD/MD_channel_0D.py (hunk 8 above)
Comment (ElmiraShamlou) [reply watertap/unit_models/MD/MD_channel_0D.py:58]: @savannahsakhai Even though the model is 0D, it intentionally includes a length domain and inlet/outlet properties, aligning with the flow direction, much like a 1D model would

**Pull request merged**